design/object-bucket/object-bucket.md

```diff
@@ -0,0 +1,300 @@
+    1. The operator gets the object store service using `storageClass.parameters.objectStoreNamespace` + `storageClass.parameters.objectStore`.
+1. The operator creates a CephObjectUser if it does not exist in order to generate the operator's access key.
+1. The operator creates the bucket using the service endpoint and access key.  The bucket name will be a concatenation of the  OBC's `metadata.namespace` and `metadata.name`.
+    1. If the bucket already exists in the store, the operator will instead regenerate the ConfigMap and Secret associated with that bucket.  This allows users the ability to reacquire buckets who's claims have been deleted.
+1. The operator creates an OB containing metadata about the new bucket and sets it as bound to the OBC.
+1. The operator updates the OBC to set it as "bound" to the OB.
+1. The operator creates a ConfigMap in the namespace of the OBC with the endpoint information for that bucket.
+    1. The operator uses the radosgw-admin CLI to generate the user's access key.  
+    1. The key is linked to the bucket.
+1. The operator creates a Secret in the namespace of the OBC with the **user's** access key.
+1. An app Pod may then reference the Secret and the ConfigMap to begin accessing the bucket.
+
+**Use Case: Delete an Object Bucket**
+
+_As a Kubernetes user, I want to delete ObjectBucketClaim instances and cleanup generated API resources._
+
+1. The user deletes the OBC via `kubectl delete objectbucketclaim -n default FOO`.
+1. The OBC is marked for deletion and left in the foreground.
+1. The respective ConfigMap and user Secret are deleted.
+1. The OBC is garbage collected.
+1. The OB phase is changed from "bound" to "lost"
+
+    Note: this does not cause the deletion of the OB.  Since the actual bucket is not deleted, it is reasonable that it should still tracked in Kubernetes.
+
+---
+
+## Looking Forward
+
+- Resource Quotas cannot yet be defined for CRDs.  This limits admin control over how many buckets can be created per object store.  A operator could be made configurable (via ConfigMap) to cap the number of buckets per object store.  A PR exists for enabling quotas on CRDs (https://github.com/kubernetes/kubernetes/pull/72384).  If it is merged into Kubernetes, a ResourceQuota example should be defined for OBCs.
+
+- Custom Bucket and Object policies are not defined for OBCs.  Currently all user keys will have Object PUT/GET/DELETE and object policy setting permissions.  It would be useful to allow users to link secondary keys with a subset of these permissions to buckets.
+
+- A generic S3 user API is not defined in this proposal but should be given serious consideration.  A user API should allow cluster users to generate new S3 users, tenants, and access keys in an object store.  This API could be used as method of manually creating object store users or incorporated into the bucket provisioning automation to generate the user's access key.
```

Comment:
Would this essentially be a general version of the CephObjectUser CRD?

Reply:
Correct.  Given that the `user` concept is ubiquitous, it makes sense that an abstract version should exist, perhaps in the same "claim" model as buckets.